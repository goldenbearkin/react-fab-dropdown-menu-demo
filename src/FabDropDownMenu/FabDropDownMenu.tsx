import * as React from 'react';
import MenuControl from './private/MenuControl';
import MenuItem from './private/MenuItem';

type PropsT = {
  density?: number,
  duration?: number
};

type StateT = {
  isOpened: boolean;
};

class FabMenu extends React.PureComponent<PropsT, StateT> {
  static defaultProps: Partial<PropsT> = {
    density: 80,
    duration: 300
  };

  constructor() {
    super();
    this.state = { isOpened: false };
    this.handleClick = this.handleClick.bind(this);
  }

  render() {
    const { duration, density, children } = this.props;
    const { isOpened } = this.state;

    const numOfChild = React.Children.count(children);
    const display = React.Children.map(children, (child, index) => {
      const subDuration = isOpened ? numOfChild - index : index + 1;
      const subDelay = !isOpened ? numOfChild - index - 1 : index;
      const inc = duration! / numOfChild;

      return (
        <div style={{ position: 'absolute', top: (index + 1) * density!, left: 0 }}>
          <MenuItem shrink={isOpened} duration={subDuration * inc} delay={subDelay * inc}>
            {child}
          </MenuItem>
        </div>
      );
    });
    return (
      <div style={{ display: 'inline-block', position: 'relative' }}>
        <MenuControl open={isOpened} onClick={this.handleClick} duration={duration!} />
        {display}
      </div>
    );
  }
  private handleClick() {
    this.setState({ isOpened: !this.state.isOpened });
  }
}

export default FabMenu;