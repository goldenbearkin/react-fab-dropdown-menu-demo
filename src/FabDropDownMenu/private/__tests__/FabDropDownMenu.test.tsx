import * as React from 'react';
import FabDropDownMenu from '../../FabDropDownMenu';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

describe('<FabDropDownMenu />', () => {
  const wrapper = shallow(<FabDropDownMenu />);
  it('should render', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });
  it('1+1=2', () => {
    expect(1 + 1).toBe(2);
  });
});