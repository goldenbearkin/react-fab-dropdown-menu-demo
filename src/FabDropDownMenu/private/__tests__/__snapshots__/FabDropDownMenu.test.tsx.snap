// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FabDropDownMenu /> should render 1`] = `
<div>
  <div
    style={
      Object {
        "display": "inline-block",
        "position": "relative",
      }
    }
  >
    <MenuControl
      backgroundColor="rgb(0, 188, 212)"
      duration={300}
      onClick={[Function]}
      open={false}
    />
  </div>
</div>
`;

exports[`rendering should render a <FabDropDownMenu /> 1`] = `
<div
  style={
    Object {
      "display": "inline-block",
      "position": "relative",
    }
  }
>
  <MenuControl
    duration={300}
    onClick={[Function]}
    open={false}
  />
</div>
`;
